### PR TITLE
Many to many relationships

### PR DESCRIPTION
**Pull request by amialkumau**
!! MERGE ONLY AFTER THIRD HOMEWORK !!

In this homework new entity `group` was added. `Groups` table was added in Postgres.
Added possibilities:
* to get group by id;
* get all groups;
* create and update group;
* remove group(hard delete);

Many-to-many relation was implemented. It was implemented by `explicit` relation in `prisma` - `UserOnGroups` table was created as link table.
* in case of group delete cascade delete will happen - all the entries associated with that group will be removed.
* in case of user delete cascade delete doesn't work because of soft delete.

Added `addUsersToGroup` endpoint, which allows to add users to specific group. It will add new entries in link table. Prisma uses transactions in case of nested writes(which we are using in that endpoint) under the hood.

There is no proper error handling ---> it will be implemented in the next homework.

### CODE REVIEW

File: Homework-4-ManyToManyRelationships/src/dao/groupDao.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
+import { PrismaClient } from ".prisma/client";
+import { Service } from "typedi";
+import type { BaseGroupDTO, GroupDTO } from "../types/groupDTO";
+
+const prisma = new PrismaClient();
```
Comment (boccob) [inline Homework-4-ManyToManyRelationships/src/dao/groupDao.ts:5]: It's recommended to have only one prisma client:
https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/instantiate-prisma-client#the-number-of-prismaclient-instances-matters

File: Homework-4-ManyToManyRelationships/src/api/groups.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+import { BaseGroupDTO, GroupDTO } from "../types/groupDTO";
+
+export const groupsRouter = express.Router();
+const groupService = Container.get(GroupService);
+
+const getAllGroups = async (_req: Request, res: Response, next: NextFunction) => {
+  try {
+    const groups = await groupService.getAllGroups();
+    res.status(200).send(groups);
+  } catch (error) {
```
Comment (boccob) [inline Homework-4-ManyToManyRelationships/src/api/groups.ts:13]: I'm wondering if this error handling is anyhow different from default one?

File: Homework-4-ManyToManyRelationships/src/api/users.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  try {
+    const filters: UserFilter = req.query;
+    const users = await userService.getUsers(filters);
+    res.status(200).send(users);
+  } catch (error: any) {
+    next(error);
+  }
+};
+
+const getUserById = async (req: Request, res: Response, next:NextFunction) => {
```
Comment (boccob) [inline Homework-4-ManyToManyRelationships/src/api/users.ts:26]: There is quite a lot of common things between api/groups and api/users. Probably not a problem with two entities. But you can have way more.

You can check `feathers` for inspiration.
The same idea could be applied to services.

**Review by boccob — APPROVED** [review]

File: Homework-4-ManyToManyRelationships/src/dao/groupDao.ts (hunk 1 above)
Comment (amialkumau) [reply Homework-4-ManyToManyRelationships/src/dao/groupDao.ts:5]: Will be fixed in next homework!

File: Homework-4-ManyToManyRelationships/src/api/groups.ts (hunk 2 above)
Comment (amialkumau) [reply Homework-4-ManyToManyRelationships/src/api/groups.ts:13]: Will make it in homework dedicated to error handling

File: Homework-4-ManyToManyRelationships/src/api/users.ts (hunk 3 above)
Comment (amialkumau) [reply Homework-4-ManyToManyRelationships/src/api/users.ts:26]: Thanks for advice

**Pull request merged**